### PR TITLE
aoi: Add way to set AOI for auto exposure.

### PR DESCRIPTION
**Pull request by jackg0**
Summary: Add way to set AOI for auto exposure.

Test plan: Test locally with camera.
  - Tested disabling AOI
  - Tested enabling AOI with correct settings
  - Tested with black object to demonstrate AOI bounds
  - Tested out of bounds AOI settings

### CODE REVIEW

File: catkin_ws/src/arena_camera/src/arena_camera_node.cpp [hunk 1]
```diff
@@ -784,6 +784,63 @@ bool ArenaCameraNode::setImageFlip(const bool flip_image)
...
+    GenApi::CBooleanPtr pAOIOffsetX = pDevice_->GetNodeMap()->GetNode("AutoExposureAOIOffsetX");
+    GenApi::CBooleanPtr pAOIOffsetY = pDevice_->GetNodeMap()->GetNode("AutoExposureAOIOffsetY");
+    if (!GenApi::IsWritable(pAOIEnable))
+    {
+      ROS_WARN_STREAM("Cannot enable AOI; cannot write to AutoExposureAOIEnable.");
+      return false;
+    }
+    Arena::SetNodeValue<bool>(pDevice_->GetNodeMap(), "AutoExposureAOIEnable", true);
+
+    if (!GenApi::IsWritable(pAOIWidth) || !GenApi::IsWritable(pAOIHeight))
```
Comment (ziparker) [inline catkin_ws/src/arena_camera/src/arena_camera_node.cpp:818]: think it'd be best to check all of these first so we don't have a case that enables aoi w/o being able to adjust it at all?
probably not a thing that can happen in the camera, but just being defensive here.

File: catkin_ws/src/arena_camera/src/arena_camera_node.cpp [hunk 2]
```diff
@@ -784,6 +784,63 @@ bool ArenaCameraNode::setImageFlip(const bool flip_image)
...
+                                                    << arena_camera_parameter_set_.aoi_offset_y_);
+  }
+  catch (const GenICam::GenericException& e)
+  {
+    ROS_ERROR_STREAM("An exception while setting AOI occurred: " << e.GetDescription());
+    return false;
+  }
+
+  return true;
+
```
Comment (ziparker) [inline catkin_ws/src/arena_camera/src/arena_camera_node.cpp]: remove extra whitespace here?

File: catkin_ws/src/arena_camera/src/arena_camera_parameter.cpp [hunk 3]
```diff
@@ -365,6 +370,18 @@ void ArenaCameraParameter::readFromRosParameterServer(const ros::NodeHandle& nh)
   nh.param<double>("exposure_search_timeout", exposure_search_timeout_, 5.);
   nh.param<double>("auto_exposure_upper_limit", auto_exp_upper_lim_, 10000000.);
 
+  if (nh.hasParam("enable_aoi"))
+  {
+    nh.getParam("enable_aoi", enable_aoi_);
+  }
```
Comment (ziparker) [inline catkin_ws/src/arena_camera/src/arena_camera_parameter.cpp]: i'd add a space between statements like this to help readability (avoid looking like else if at a glance)

File: catkin_ws/src/arena_camera/src/arena_camera_parameter.cpp [hunk 4]
```diff
@@ -365,6 +370,18 @@ void ArenaCameraParameter::readFromRosParameterServer(const ros::NodeHandle& nh)
   nh.param<double>("exposure_search_timeout", exposure_search_timeout_, 5.);
   nh.param<double>("auto_exposure_upper_limit", auto_exp_upper_lim_, 10000000.);
 
+  if (nh.hasParam("enable_aoi"))
```
Comment (ziparker) [inline catkin_ws/src/arena_camera/src/arena_camera_parameter.cpp]: do we need to check hasParam first?

File: catkin_ws/src/arena_camera/src/arena_camera_node.cpp [hunk 5]
```diff
@@ -784,6 +784,63 @@ bool ArenaCameraNode::setImageFlip(const bool flip_image)
...
+      return false;
+    }
+    Arena::SetNodeValue<bool>(pDevice_->GetNodeMap(), "AutoExposureAOIEnable", true);
+
+    if (!GenApi::IsWritable(pAOIWidth) || !GenApi::IsWritable(pAOIHeight))
+    {
+      ROS_WARN_STREAM("Cannot configure AOI size; cannot write to AutoExposureAOIHeight or Width.");
+      return false;
+    }
+    Arena::SetNodeValue<int64_t>(pDevice_->GetNodeMap(),
```
Comment (ziparker) [inline catkin_ws/src/arena_camera/src/arena_camera_node.cpp:833]: just checking - these params in the node map are all 64 bit ints?

Comment (jackg0) [reply catkin_ws/src/arena_camera/src/arena_camera_node.cpp:833]: Yeah, IInteger interface on GenICam uses 64-bit signed ints.

**Review by ziparker — APPROVED** [review]

**Pull request merged**